### DISCUSSION TITLE
Are nested objects supported in a lens selector?

### DISCUSSION THREAD

**Discussion by zkxs** [yamafaktory/jql#224] [Q&A] [answered]
Hi, I may be missing something obvious, but I can't figure out if the lens selector can filter against nested objects or not.

Here's an example input json:

```json
[
    {
        "metadata": {
            "unique_id": "fizzbuzz"
        },
        "data": ["foo", "bar", "baz"]
    },
    {
        "metadata": {
            "unique_id": "giant"
        },
        "data": ["fee", "fi", "fo", "fum"]
    }
]
```

What I'd like to do is get the `data` array for all objects with a specific `unique_id`. 

So if I'm looking for `"unique_id"="fizzbuzz"`, then I'd expect the following output:

```json
[
  [
    "foo",
    "bar",
    "baz"
  ]
]
```

Intuitively, I thought the lens selector would be able to do something like this, but in practice it gives a parse error:
```bash
jql '|={"metadata""unique_id"="fizzbuzz"}' example.json # Unable to parse input
```

The only thing I was able to figure out was a workaround with flattening. While this does give the desired output, flattening is destructive and I'd rather avoid it if possible.

```bash
jql '|> .. <| |={"metadata.unique_id"="fizzbuzz"} |> "data" <|' example.json # desired output
```

**Comment by yamafaktory** [comment, marked as answer]
Hi @zkxs,

Thanks for your feedback!

Indeed this is not possible in the current state of `jql` since a `Lens` is basically a tuple of key (`str`) + `LensValue` (a JSON primitive):

https://github.com/yamafaktory/jql/blob/dbf899768686d500dec14b700b0a45cf785c5bac/crates/jql-parser/src/tokens.rs#L75

https://github.com/yamafaktory/jql/blob/dbf899768686d500dec14b700b0a45cf785c5bac/crates/jql-parser/src/tokens.rs#L109-L118

To make it work, I'll need to refactor the parser implementation to allow a `Lens` to hold other selectors as __key__. 

I'll think about it and see if that's not going to add too much complexity.

**Comment by yamafaktory** [comment]
@zkxs I've prepared a draft PR to allow a subset of selectors to work with lenses. I need to do some more testing (and update the documentation too) and then will make a new release.

**Comment by yamafaktory** [comment]
@zkxs I've just released a new major version which now supports lenses with selectors. Thanks a lot for bringing this issue to the table 👍🏼 .
